### PR TITLE
Videopress Block: Fix video width and lack of alignment classes

### LINKED ISSUE #17925: VideoPress: Video blocks encounter errors when using VideoPress URL
#### Steps to reproduce the issue

1. Upload a video so it's on VideoPress
2. Insert a new column block or layout grid
3. Insert a video block and add the VideoPress URL
4. Save and refresh
5. Video will show an error.

#### What I expected

I expect the video to keep working. Although this isn't a standard way of adding videos it is quite easy to do it this way and end up with errors in blocks that can't be easily fixed.

#### What happened instead
The blocks break unless the video block contains a link to the video file directly in the media library instead of the VideoPress URL.

Replicated on a WordPress.com Business plan site and a WordPress.org site with Gutenberg plugin 9.3 and 9.4.1 both causing this error. It happened with and without the Layout Grid.

**Screenshots**
Markup 2020-11-24 at 17 55 45

### LINKED ISSUE #46889: Video Block: Width setting doesn't work
In WP.com, the width setting for Video block doesn't work on the front of the site, and it doesn't seem to be a theme issue.

screencapture-iamtakashi-wordpress-2020-10-28-headings-on-cover-2020-10-28-20_54_32 1

In the editor, they appear as expected.

cc @ianstewart @apeatling

### PR DESCRIPTION
**Pull request by aaronrobertshaw**
Fixes https://github.com/Automattic/wp-calypso/issues/46889
Fixes #17925 

VideoPress block on WordPress.com doesn't get the alignment CSS classes applied to it on save. This leads to the video appearing to be different widths in editor vs frontend.

#### Changes proposed in this Pull Request:
* Update application of CSS class names it to pick up the alignment CSS classes.
* Fix block attributes to prevent block validation errors that were occurring in production
* Move setting of `classNames` attribute out of the `render()` function to avoid `UnsavedChangesWarning` error.
* Add deprecation for older blocks that were missing alignment classes or didn't get `classNames` saved due to bug.

#### Does this pull request change what data or activity we track or use?
No.

#### Unit test for attribute migrations
1. Run `yarn test-extensions`
2. Confirm `extensions/blocks/videopress/test/index.js` ran successfully.

#### Testing instructions - Local:

1. Using master branch, start up local Jetpack site with sufficient plan to upload videos to videopress
2. Add a video block to a post and upload a video
3. Set it's alignment to `wide` and add a custom CSS class name via sidebar > advanced controls
4. Save post and view on frontend. (Should be missing `alignwide` CSS class)
5. Reload post in the editor, the block's will be invalid due to current bugs.
5. Checkout this PR and rebuild Jetpack
6. With dev tools console open, reload the editor and confirm no errors and the block deprecation is successfully applied
7. Save the post and reload the frontend. 
8. Confirm the block displays with wide alignment

#### Testing instructions - Sandbox:

1. Apply this PR branch and test VideoPress block functions correctly locally.
2. On a sandboxed site with a sufficient plan, add a video to a post, set its alignment to "wide", then save.
3. Confirm on the frontend that the video's block does not contain the `alignwide` CSS class.
4. Apply this PR's associated diff ( D53903-code ) or use a8c-wp-env and sync Jetpack to your sandbox.
5. Reload the editor for your sandboxed site and confirm the block is still valid.
6. Change the block's alignment to "full" and save post.
7. Switch to the frontend, reload and confirm the video block has the `alignfull` CSS class.

#### Screenshots

| Wide Align - Before | Wide Align - After |
|------------------------|-----------------------|
| Screen Shot 2020-12-08 at 6 32 41 pm | Screen Shot 2020-12-08 at 6 33 42 pm |

#### Proposed changelog entry for your changes:
* Fixes missing alignment classes on VideoPress block.
* Fixes VideoPress block validation errors

### CODE REVIEW

**Comment by glendaviesnz** [comment]
I think we might not be able to use `useBlockProps` as it won't be supported by all the versions of the editor that jetpack supports - looks like it was only renamed and stabalised 2 months ago - https://github.com/WordPress/gutenberg/pull/25642

**Review by apeatling — CHANGES_REQUESTED** [review]
We chatted about this, but I think you could check for the presence of `useblockProps` and fall back to manual if not present. Or go with manual until Jetpack catches up to this.

**Comment by roundhill** [comment]
Stumbled across this PR, thanks for fixing @aaronrobertshaw 😄

**Comment by aaronrobertshaw** [comment]
This has been updated to check for `useBlockProps` as well as being rebased.

I've tested this using the latest Gutenberg release and [v9.1.1](https://github.com/WordPress/gutenberg/releases/tag/v9.1.1) just prior to when commit https://github.com/WordPress/gutenberg/pull/25642 stabilised the block API involving `useBlockProps` in [v9.2.0](https://github.com/WordPress/gutenberg/releases/tag/v9.2.0)

**Review by apeatling — DISMISSED** [review]

**Review by jeherve — CHANGES_REQUESTED** [review]
Since we're changing the saved markup of our blocks, we'll need to handle block deprecation to avoid block validation errors when editing existing posts. Could you add this?

**Review by apeatling — DISMISSED** [review]

**Comment by apeatling** [comment]
@jeherve Does this one look GTM now?

**Review by jeherve — COMMENTED** [review]
This doesn't seem to work for me in Jetpack (no Gutenberg).

I'm still getting block validation errors when trying to edit posts, and I do not see any alignment classes added:

```
blocks.js:8921 Block validation: Block validation failed for `core/video` (Object).

Content generated by `save` function:

<figure class="wp-block-video wp-block-embed is-type-video is-provider-videopress wp-embed-aspect-4-3 wp-has-aspect-ratio"><div class="wp-block-embed__wrapper">
https://videopress.com/v/Uvh9ON4T?preloadContent=metadata
</div></figure>

Content retrieved from post body:

<figure class="wp-block-embed is-type-video is-provider-videopress wp-embed-aspect-4-3 wp-has-aspect-ratio"><div class="wp-block-embed__wrapper">
https://videopress.com/v/Uvh9ON4T?preloadContent=metadata
</div></figure>
```

**Comment by aaronrobertshaw** [comment]
> I'm still getting block validation errors when trying to edit posts, and I do not see any alignment classes added:

This one had me scratching my head. 

It turns out it's broken on master and WordPress.com. A few months ago the block was updated to set a new `classNames` attribute without actually defining it in the block's attributes. Compounding this, the change was also attempting to set the attribute within the render function resulting in another error.

I've rebased this PR and updated it to fix the above issues as well. I no longer receive any errors while testing this though it would be great if the approach I have taken can be double checked with fresh eyes.

**Review by apeatling — DISMISSED** [review]
Retested this and it is working for me, not seeing the above errors. :+1: Edit: this was just on 5.6.

**Comment by jgcaruso** [comment]
I'm running into the same issue after upgrading my local Jetpack dev docker instance (and also on a clean Docker instance). I'm wondering if that error has anything to do with the change made to the core video block recently. It might only be happening if you are using the latest version of WordPress?

This change was made 28 days ago to the core video block, and it matches the extra class name that @jeherve had in his error message https://github.com/WordPress/gutenberg/commit/37163dcba1a2720522681fef5d258231d7890ea6#diff-8f07e0f9951e498bb23361a304adfed92d77be5fed3fa3dd1a9b99096d4e27a2R77

Any change you've been developing against an older WP version?

**Comment by roundhill** [comment]
I also saw the error on 5.5, and updating to 5.6 seems to have resolved it while running this branch.

**Comment by aaronrobertshaw** [comment]
I didn't get the errors when testing with 5.5.3 (either on master or this PR branch). 

Testing with 5.6 however, master does throw the errors. With this PR applied, I do not see any errors.

**Review by jeherve — COMMENTED** [review]
I'm still experiencing issues with the Video block, I don't think we can merge this as it is. Here is what I tried:

- Jetpack + WP 5.5.3
- Jetpack + WP 5.6
- Jetpack + WP 5.6 + Gutenberg 9.7.0

For each environment, I used Twenty Twenty One. I created 2 posts, while on Jetpack Stable:

- One post with a video block with no custom alignment set up.
- One post with a video block with alignment set to Wide.

I then switched to this branch, and viewed the posts on the frontend, then opened them in the editor. Here are my results:

**Jetpack + WP 5.5.3**

- Video with no custom alignment, frontend: ✅ 
- Video with no custom alignment, editor: ✅ 
- Video with wide align, frontend: ✅ 
- Video with wide align, editor: ✅ 

**Jetpack + WP 5.6**

- Video with no custom alignment, frontend: ✅ 
- Video with no custom alignment, editor: validation error
image

- Video with wide align, frontend: the custom alignment is nowhere to be found:
image

- Video with wide align, editor: validation error
image

**Jetpack + WP 5.6 + Gutenberg 9.7.0**

- Video with no custom alignment, frontend: ✅ 
- Video with no custom alignment, editor: validation error 

image

- Video with wide align, frontend: wide alignment is not respected:
image

- Video with wide align, editor: validation error
image

**Comment by aaronrobertshaw** [comment]
@jeherve Thanks for the extra details. That is different that what I experienced earlier today. I'll investigate further.

Did you happen to reload the post in the editor at all after initially creating them while still on Jetpack master? Just looking to confirm you also got the same validation errors at that stage.

**Comment by jeherve** [comment]
> Did you happen to reload the post in the editor at all after initially creating them while still on Jetpack master? Just looking to confirm you also got the same validation errors at that stage.

The validation errors do happen on `master` as well, indeed.

**Comment by aaronrobertshaw** [comment]
I've rebased this PR and updated it to address the issue where newer versions of Gutenberg caused all the deprecations to be invalid. Tested with WordPress 5.5.3 and 5.6 as well as latest Gutenberg.

The issue with the `wp-block-video` class being added via the `blocks.getSaveContent.extraProps` filter was solved using another filter targeting deprecated blocks and adjusting the results so that the original saved content for a deprecated block would be successfully matched.

If anyone has any suggestions as to a better approach, I'd be happy to hear them.

File: extensions/blocks/videopress/deprecated/v2/index.js [hunk 1]
```diff
@@ -59,17 +59,16 @@ export default {
...
 
-		return guid && ( missingClassNames || missingVideoClass );
+		return guid && missingClassNames;
 	},
 	migrate: attributes => {
 		const { align, className, videoPressClassNames } = attributes;
 		return {
 			...attributes,
-			className: 'wp-block-video',
+			className: '',
```
Comment (glendaviesnz) [inline extensions/blocks/videopress/deprecated/v2/index.js:71]: We will lose any custom classes a user added to block doing this, so probably need to add here some way of removing all the videopress classes from className at this point and just leave custom ones

Comment (aaronrobertshaw) [reply extensions/blocks/videopress/deprecated/v2/index.js:71]: They would have been included into the `videoPressClassNames` attribute. That's how the block was previously working.

Comment (aaronrobertshaw) [reply extensions/blocks/videopress/deprecated/v2/index.js:71]: Actually, after switching back to master and testing, master loses custom CSS classes.

**Comment by glendaviesnz** [comment]
This tests well for me now, and the migrations run correctly for the previous versions.

File: extensions/blocks/videopress/deprecated/v2/index.js (hunk 1 above)
Comment (aaronrobertshaw) [reply extensions/blocks/videopress/deprecated/v2/index.js]: The migration of `className` to `videoPressClassNames` now takes into account custom CSS classes.

**Review by jeherve — APPROVED** [review]
This works well for me now; it should be good to merge!

**Comment by jeherve** [comment]
r220372-wpcom

**Pull request merged**